quil-rs/src/instruction/extern_call.rs

```diff
@@ -0,0 +1,2060 @@
+    /// names, uniqueness, and the presence of return or parameters in the signature.
+    pub(crate) fn validate_all(
+        extern_definitions: &[ExternDefinition],
+    ) -> Result<(), ExternValidationError> {
+        extern_definitions
+            .iter()
+            .try_fold(HashSet::new(), |mut acc, extern_definition| {
+                validate_user_identifier(extern_definition.name.as_str())
+                    .map_err(ExternValidationError::Name)?;
+                if acc.contains(&extern_definition.name) {
+                    return Err(ExternValidationError::Duplicate(
+                        extern_definition.name.clone(),
+                    ));
+                }
+                if let Some(signature) = extern_definition.signature.as_ref() {
+                    if !signature.has_return_or_parameters() {
+                        return Err(ExternValidationError::NoReturnOrParameters(
+                            extern_definition.name.clone(),
+                        ));
+                    }
+                }
+                acc.insert(extern_definition.name.clone());
+                Ok(acc)
+            })?;
+        Ok(())
+    }
+}
+
+/// An error that can occur when resolving a call instruction.
+#[derive(Clone, Debug, thiserror::Error, PartialEq)]
+pub enum CallArgumentResolutionError {
+    /// An undeclared memory reference was encountered.
+    #[error("undeclared memory reference {0}")]
+    UndeclaredMemoryReference(String),
+    /// A mismatched vector was encountered.
+    #[error("mismatched vector: expected {expected:?}, found {found:?}")]
+    MismatchedVector { expected: Vector, found: Vector },
+    /// A mismatched scalar was encountered.
+    #[error("mismatched scalar: expected {expected:?}, found {found:?}")]
+    MismatchedScalar {
+        expected: ScalarType,
+        found: ScalarType,
+    },
+    /// The argument for a vector parameter was invalid.
+    #[error("vector parameters must be passed as an identifier, found {0:?}")]
+    InvalidVectorArgument(UnresolvedCallArgument),
+    /// the argument for a return parameter was invalid.
+    #[error("return argument must be a memory reference, found {found:?}")]
+    ReturnArgument { found: UnresolvedCallArgument },
+}
+
+/// A parsed, but unresolved call argument. This may be resolved into a [`ResolvedCallArgument`]
+/// with the appropriate [`ExternSignature`]. Resolution is required for building the
+/// [`crate::Program`] memory graph.
+#[derive(Clone, Debug, PartialEq)]
+pub enum UnresolvedCallArgument {
+    /// A reference to a declared memory location. Note, this may be resolved to either
+    /// a scalar or vector. In the former case, the assumed index is 0.
+    Identifier(String),
+    /// A reference to a memory location. This may be resolved to a scalar.
+    MemoryReference(MemoryReference),
+    /// An immediate value. This may be resolved to a scalar.
+    Immediate(Complex64),
```

Comment:
```suggestion
    /// A reference to a memory location. This can only be resolved to a scalar.
    MemoryReference(MemoryReference),
    /// An immediate value. This can only be resolved to a scalar.
    Immediate(Complex64),
```